Interact with contracts: upload, instantiate and call commands

ascjones opened the pull request:
Adds `upload`, `instantiate` and `call` commands which allows interacting with contracts on live chains with a compatible `pallet-contracts`.

## Upload

```bash
cargo contract upload --suri //Alice --dry-run
```

![image](https://user-images.githubusercontent.com/75586/153376009-e5b1f9f1-9651-48d2-9133-f6c6702896a9.png)

```bash
cargo contract upload --suri //Alice
```

![image](https://user-images.githubusercontent.com/75586/153375560-c2111d04-9b84-4a76-8e79-a25db3603d4b.png)

## Instantiate

```bash
cargo contract instantiate \
       --constructor new \
       --args false \
       --suri //Alice \
       --code-hash 0xbc1b42256696c8a4187ec3ed79fc602789fc11287c4c30926f5e31ed8169574e \
       --dry-run

```

![image](https://user-images.githubusercontent.com/75586/153376420-ada64652-7c33-4d48-8646-b72ca27e32f0.png)

```bash
cargo contract instantiate \
       --constructor new \
       --args false \
       --suri //Alice \
       --code-hash 0xbc1b42256696c8a4187ec3ed79fc602789fc11287c4c30926f5e31ed8169574e
```

![image](https://user-images.githubusercontent.com/75586/153377038-5b291105-7686-4019-9686-4f803a9eba6d.png)

### Instantiate with code

Omitting `--code-hash` will call `instantiate_with_code`:

![image](https://user-images.githubusercontent.com/75586/153380215-f7f3ba69-8376-4439-82c8-cb3e6e4e62ab.png)

## Call

```bash
cargo contract call \
       --contract 5FQNUFjuSxjB1o67CneHXieB1Hy6xAj2A4h4dUVLCvNkgRmD \
       --message flip \
       --suri //Alice \
       --dry-run
```

![image](https://user-images.githubusercontent.com/75586/153378225-b207c2cb-18fe-4d74-91ad-23e58a5df3e7.png)

```bash
cargo contract call \
       --contract 5FQNUFjuSxjB1o67CneHXieB1Hy6xAj2A4h4dUVLCvNkgRmD \
       --message flip \
       --suri //Alice
```

![image](https://user-images.githubusercontent.com/75586/153378714-6fcfc893-bc30-42fa-a330-69524da6ac0c.png)

```sh
cargo contract call \
       --contract 5FQNUFjuSxjB1o67CneHXieB1Hy6xAj2A4h4dUVLCvNkgRmD \
       --message get \
       --suri //Alice \
       --dry-run
```

### Debug message

If the contract is build in debug mode then 

![image](https://user-images.githubusercontent.com/75586/153380474-f3caf1fe-8156-4719-95b0-98a36ef8ca84.png)

Will result in...

![image](https://user-images.githubusercontent.com/75586/153380662-dbb255a4-2a6b-404f-b122-e19ffd8fe08f.png)

# Todo

- [x] Display all events resulting from `Call`
- [x] Optional `0x` prefix for hex encoded code hash
- [x] SS58 AccountIds
- [x] Release and use `pallet-contracts-primitives` and `sp-rpc` from `crates.io`
- [x] Documentation
  - [x] CLI usage help
  - [x] extrinsics.md
  - [x] transcoding/scon code docs 

# Follow ups

These will be promoted to their own issue, or into separate issues for following up once this is merged.

- Signing: safer alternatives for value bearing chains than providing secret key on the command line. Possibly allow using encrypted file with private key, or integrating with external signers somehow.
- Dynamic `DispatchError::Module` errors, so that error details are displayed correctly across different chains https://github.com/paritytech/subxt/issues/443
- Gas estimation: auto dry run first to use estimated gas as a value for the gas limit when running actual call.
- Add flag `--wait-for-finalization`. so that success/events are not displayed until the block is finalized
- Denominated units for Balances e.g. `100CAN`, `1MCAN`
- Bitvec encoding/decoding
- Improve error message when failing to encode contract message arg https://github.com/paritytech/cargo-contract/pull/79/commits/e40dc8051b37074afad69f18ccdc35e1de408dd8
-  Set up CI with `substrate-contracts-node` and unignore integration test: https://github.com/paritytech/cargo-contract/blob/60f65a2863be1266882bdd81861d836baeb06724/src/cmd/extrinsics/integration_tests.rs#L124
- Add support for remove_code extrinsic.

cmichi commented on src/cmd/extrinsics/call.rs:
```diff
@@ -0,0 +1,145 @@
+    /// The value to be transferred as part of the call
+    #[structopt(name = "value", long, default_value = "0")]
+    value: <ContractsTemplateRuntime as Balances>::Balance,
+    #[structopt(name = "contract", long)]
```
```suggestion
    #[structopt(name = "contract", long, env = "CONTRACT")]
```

This will use the environment variable `CONTRACT` as a fallback if a) it exists and  b) `--contract` was not given on the command line – thus enabling kind of a "session mode" :-).

ascjones replied:
TIL :rocket:

cmichi commented on src/cmd/extrinsics/transcode/encode.rs:
```diff
@@ -0,0 +1,330 @@
+                    u.encode_to(output);
+                    Ok(())
+                } else {
+                    Err(anyhow::anyhow!("Expected a u16 value"))
```
```suggestion
                    Err(anyhow::anyhow!("Expected a u32 value"))
```

ascjones commented:
## Status update

On the backburner while I work on integrating scale-info into substrate, which will benefit this feature by reducing large compile times by removing substrate dependencies (via subxt). Also it will be possible to target different chains and different versions of the contracts pallet.

brson commented:
Epic pr. Not sure why I'm getting notifications on it, but good work @ascjones.

athei commented:
Would be great if the RPCs could be moved to `subxt` so that they can be used by other crates, too. In general, it would be best if everything needed to interact with a node would be either implemented or re-exported by `subxt` so this would be the go-to dependency to do so.

ascjones commented on Cargo.toml:
```diff
@@ -43,14 +43,25 @@ url = { version = "2.2.2", features = ["serde"] }
+sp-core = "4.0.0"
+sp-keyring = "4.0.0"
+sp-runtime = "4.0.0"
+pallet-contracts-primitives = { package = "pallet-contracts-primitives", git = "https://github.com/paritytech/substrate/", branch = "master" }
```
Now other `sp-*` crates are released, we need to release `pallet-contracts-primitives`, and also `sp-rpc` on which it depends. /cc @athei

athei replied:
We don't really do releases. Can't you stick with a git ref for everything?

ascjones replied:
Not if we want to release `cargo-contract` to `crates.io`. But that is not an absolute requirement for now.

The alternative would be to copy those types we need over here, or to implement https://github.com/paritytech/subxt/issues/416

athei replied:
Released the crate. You can switch to the released version now. Also make sure to bump the rest of the substrate versions.

ascjones replied:
Thank you very much :pray: 

We also need a `5.0.0` of `sp-keyring`: https://github.com/paritytech/substrate/pull/10790

ascjones commented on Cargo.toml:
```diff
@@ -28,7 +28,7 @@ heck = "0.4.0"
 parity-wasm = "0.42.2"
 cargo_metadata = "0.14.1"
-codec = { package = "parity-scale-codec", version = "2.3.1", features = ["derive"] }
+scale = { package = "parity-scale-codec", version = "2.3.1", features = ["derive"] }
```
Renamed because I have an embedded contract for testing purposes and `ink!` generates code with `::scale` imports.

athei commented on Cargo.toml:
```diff
@@ -43,14 +43,26 @@ url = { version = "2.2.2", features = ["serde"] }
+ink_metadata = { version = "3.0.0-rc8", features = ["derive"] }
+ink_env = "3.0.0-rc8"
+scale-info = { version = "1.0.0", default-features = false, features = ["derive"] }
+sp-arithmetic = "4.0.0"
+sp-core = "5.0.0"
+sp-keyring = "4.0.0"
```
Those pulled in a lot of duplicated dependencies and are actually unused in the code.

```suggestion
sp-core = "5.0.0"
```

athei commented on docs/extrinsics.md:
````diff
@@ -0,0 +1,103 @@
+*Optional*. The password for the `--suri`, see https://docs.substrate.io/v3/tools/subkey/#password-protected-keys.
+
+```
+--manfest-path
````
```suggestion
--manifest-path
```

athei reviewed (approved):
I played around with it a bit and it is working as I imagined it. Some remarks:

### Probably fix now
* Remove trailing new line from debug message. It looks awkward with the quote on the next line.
* Remove the decoding of the module error as long as this is buggy. It is massively confusing to see a wrong error there. Better to not decode.


### Fix in follow up
* Add support for `remove_code` extrinsic.

ascjones commented:
>     * Remove trailing new line from debug message. It looks awkward with the quote on the next line.

Changed this so that it will print successive lines indented:

![image](https://user-images.githubusercontent.com/75586/153952167-b3208923-c868-455c-aef3-3ca3950757b1.png)

Other issues addressed and added follow up to list in PR.

cmichi commented on Cargo.toml:
```diff
@@ -43,14 +43,24 @@ url = { version = "2.2.2", features = ["serde"] }
+async-std = { version = "1.10.0", features = ["attributes", "tokio1"] }
+ink_metadata = { version = "3.0.0-rc8", features = ["derive"] }
+ink_env = "3.0.0-rc8"
+scale-info = { version = "1.0.0", default-features = false, features = ["derive"] }
```
This will probably trigger a dependabot PR for an upgrade to `2.0.0`.

ascjones replied:
I'll keep it as `1.0.0` for now since that is the same version as `ink_metadata` `rc8` is using.
